proxy/src/connection.rs

```diff
@@ -186,6 +189,61 @@ impl BoundPort {
     }
 }
 
+// ===== impl ConditionallyUpgradeServerToTls =====
+
+impl ConditionallyUpgradeServerToTls {
+    fn new(socket: TcpStream, tls: tls::ConnectionConfig<tls::ServerConfig>) -> Self {
+        ConditionallyUpgradeServerToTls::Plaintext(Some(ConditionallyUpgradeServerToTlsInner {
+            socket,
+            tls,
+            peek_buf: BytesMut::with_capacity(8192),
+        }))
+    }
+}
+
+impl Future for ConditionallyUpgradeServerToTls {
+    type Item = Connection;
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        loop {
+            *self = match self {
+                ConditionallyUpgradeServerToTls::Plaintext(ref mut inner) => {
+                    let r = {
+                        let inner = inner.as_mut().unwrap();
+                        try_ready!(inner.socket.read_buf(&mut inner.peek_buf));
```

Comment:
Just as a refresher, how many bytes would we expect to need to read before knowing where to upgrade or not? `BytesMut` won't reserve more space automatically, is the original 8192 bytes sufficient?

Reply:
We would need way less than 8192. I have a branch around somewhere to decrease this size, but  I kept it that size since I didn't know how it would impact the current (HTTP) protocol disambiguation.

Also, I realize that this is pretty much the same as the current (HTTP) protocol disambiguation except that this TLS disambiguation is in two distinct phases: protocol disambiguation at the start of a connection followed by the creation of a unambiguous channel. If the HTTP protocol disambiguation can be cast that way (i.e. if we don't need to do peeking except at the start of a connection) then I think it would be good to change the HTTP disambiguation to be more like this and (I hope) unify the two.

Reply:
To detect HTTP2, the bytes must start with exactly `PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n`. It's probably likely that a similar size string that otherwise looks like HTTP/1.x (such that, `httparse` returns that it is `Partial` and not an `Error`).